### PR TITLE
fix:  improve message for inactive weak optional feature with edition2024

### LINKED ISSUE #14015: Bad error message when a weak dep features's depedency is an unused optional dependency
### Problem

If nothing can activate an optional dependency, cargo acts as if the dependency doesn't exist which creates poor error messages.

Also, with how things are arranged, the unused optional dependency lint doesn't get reported which could at least reduce the burden on the error message.

### Steps

Baseline `Cargo.toml`:
```toml
cargo-features = ["edition2024"]
[package]
name = "cargo-14010"
version = "0.1.0"
edition = "2024"

[dependencies]
serde = { version = "1.0.203", optional = true }
```
```console
$ cargo +nightly check -Zcargo-lints
warning: unused optional dependency
 --> Cargo.toml:8:1
  |
8 | serde = { version = "1.0.203", optional = true }
  | -----
  |
  = note: `cargo::unused_optional_dependency` is set to `warn` by default
  = help: remove the dependency or activate it in a feature with `dep:serde`
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.00s
```

Add the following to `Cargo.toml`:
```toml

[features]
serde = ["serde?/derive"]
```
```console
$ cargo +nightly check -Zcargo-lints
error: failed to parse manifest at `/home/epage/src/personal/dump/cargo-14010-1/Cargo.toml`

Caused by:
  feature `serde` includes `serde?/derive`, but `serde` is not a dependency
```

### Possible Solution(s)

- Improve the error message
- Show the lint (which might be more difficult)

### Notes

_No response_

### Version

_No response_

### PR DESCRIPTION
**Pull request by linyihai**
### What does this PR try to resolve?

This will improve the message for inactive weak optional feature with edition2024.

This doesn't distinguish whether the dependency had set, which needs to get the dependency from origin that will make more complex.

Fixes #14015 

### How should we test and review this PR?

one commit add the test, one commit fix and update the test

### CODE REVIEW

File: tests/testsuite/lints/unused_optional_dependencies.rs [hunk 1]
```diff
@@ -221,7 +221,7 @@ Caused by:
 error: failed to parse manifest at `[ROOT]/foo/Cargo.toml`
 
 Caused by:
-  feature `feat` includes `dep?/feat`, but `dep` is not a dependency
+  feature `feat` includes `dep?/feat`, activate it in a feature with `dep:dep` if `dep` is an enabled dependency
```
Comment (epage) [inline tests/testsuite/lints/unused_optional_dependencies.rs:224]: > if `dep` is an enabled dependency

I feel the wording could be improved here.  To help in coming up with ideas, could you expand on what you are trying to convey

Comment (linyihai) [reply tests/testsuite/lints/unused_optional_dependencies.rs:224]: Yes, the messge need to include the cases that dependency inactive and dependency missing, `if dep is an enabled dependency` means the former.

https://github.com/rust-lang/cargo/pull/14026 is another PR to distinguish these. And this PR may require more detailed wording to address both cases.

Comment (linyihai) [reply tests/testsuite/lints/unused_optional_dependencies.rs:224]: I suggest keep this PR to polish the message unless https://github.com/rust-lang/cargo/pull/14026https://github.com/rust-lang/cargo/pull/14026 is merged

**Comment by linyihai** [comment]
Close this in favor of https://github.com/rust-lang/cargo/pull/14026

**Pull request closed without merging**